### PR TITLE
Messaging updates

### PR DESCRIPTION
**Pull request by hananiel**

### CODE REVIEW

File: src/Integration/src/IntegrationBase/IntegrationServices.cs [hunk 1]
```diff
@@ -32,7 +36,11 @@ public virtual IMessageBuilderFactory MessageBuilderFactory
             {
                 if (_messageBuilderFactory == null)
                 {
-                    _messageBuilderFactory = _context.GetService<IMessageBuilderFactory>();
+                    _messageBuilderFactory = _context?.GetService<IMessageBuilderFactory>();
+                    if (_messageBuilderFactory == null)
+                    {
+                        _messageBuilderFactory = new DefaultMessageBuilderFactory();
+                    }
```
Comment (TimHess) [inline src/Integration/src/IntegrationBase/IntegrationServices.cs:43]: It looks like there are several instances of this pattern in a row, I'll just make this suggestion once to get your reaction: 
```suggestion
                    _messageBuilderFactory = _context?.GetService<IMessageBuilderFactory>() ?? new DefaultMessageBuilderFactory();
```

File: src/Integration/src/IntegrationBase/Transformer/JsonToObjectTransformer.cs [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+                result = JsonConvert.DeserializeObject((string)payload, targetClass, Settings);
+            }
+            else if (payload is byte[])
+            {
+                var contentAsString = DefaultCharset.GetString((byte[])payload);
+                result = JsonConvert.DeserializeObject(contentAsString, targetClass, Settings);
+            }
+            else
+            {
+                throw new MessageConversionException("Failed to convert Message content, message missing byte[] or string:" + payload.GetType());
```
Comment (TimHess) [inline src/Integration/src/IntegrationBase/Transformer/JsonToObjectTransformer.cs:77]:
```suggestion
                throw new MessageConversionException("Failed to convert Message content, message missing byte[] or string: " + payload.GetType());
```

File: src/Integration/src/IntegrationBase/IntegrationServices.cs [hunk 3]
```diff
@@ -44,13 +48,39 @@ public virtual IMessageBuilderFactory MessageBuilderFactory
...
+        {
+            get
+            {
+                if (_expressionParser == null)
+                {
+                    _expressionParser = _context?.GetService<IExpressionParser>();
+                    if (_expressionParser == null)
+                    {
+                        _expressionParser = new ExpressionParser();
+                    }
```
Comment (TimHess) [inline src/Integration/src/IntegrationBase/IntegrationServices.cs:61]:
```suggestion
                    _expressionParser = _context?.GetService<IExpressionParser>() ?? new ExpressionParser();
```

File: src/Integration/src/IntegrationBase/IntegrationServices.cs [hunk 4]
```diff
@@ -44,13 +48,39 @@ public virtual IMessageBuilderFactory MessageBuilderFactory
...
             get
             {
                 if (_channelResolver == null)
                 {
-                    _channelResolver = _context.GetService<IDestinationResolver<IMessageChannel>>();
+                    _channelResolver = _context?.GetService<IDestinationResolver<IMessageChannel>>();
+                    if (_channelResolver == null)
+                    {
+                        _channelResolver = new DefaultMessageChannelResolver(_context);
+                    }
```
Comment (TimHess) [inline src/Integration/src/IntegrationBase/IntegrationServices.cs:83]:
```suggestion
                    _channelResolver = _context?.GetService<IDestinationResolver<IMessageChannel>>() ?? new DefaultMessageChannelResolver(_context);
```

File: src/Integration/src/IntegrationBase/IntegrationServices.cs [hunk 5]
```diff
@@ -68,7 +98,11 @@ public virtual IConversionService ConversionService
             {
                 if (_conversionService == null)
                 {
-                    _conversionService = _context.GetService<IConversionService>();
+                    _conversionService = _context?.GetService<IConversionService>();
+                    if (_conversionService == null)
+                    {
+                        _conversionService = DefaultConversionService.Singleton;
+                    }
```
Comment (TimHess) [inline src/Integration/src/IntegrationBase/IntegrationServices.cs:105]:
```suggestion
                    _conversionService = _context?.GetService<IConversionService>() ?? DefaultConversionService.Singleton;
```

File: src/Integration/src/IntegrationBase/IntegrationServices.cs [hunk 6]
```diff
@@ -86,7 +120,11 @@ public virtual IIDGenerator IdGenerator
             {
                 if (_idGenerator == null)
                 {
-                    _idGenerator = _context.GetService<IIDGenerator>();
+                    _idGenerator = _context?.GetService<IIDGenerator>();
+                    if (_idGenerator == null)
+                    {
+                        _idGenerator = new DefaultIdGenerator();
+                    }
```
Comment (TimHess) [inline src/Integration/src/IntegrationBase/IntegrationServices.cs:127]:
```suggestion
                    _idGenerator = _context?.GetService<IIDGenerator>() ?? new DefaultIdGenerator();
```

File: src/Integration/src/IntegrationBase/Transformer/AbstractTransformer.cs [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+        {
+            get
+            {
+                if (_messageBuilderFactory == null)
+                {
+                    _messageBuilderFactory = ApplicationContext.GetService<IMessageBuilderFactory>();
+                    if (_messageBuilderFactory == null)
+                    {
+                        _messageBuilderFactory = new DefaultMessageBuilderFactory();
+                    }
```
Comment (TimHess) [inline src/Integration/src/IntegrationBase/Transformer/AbstractTransformer.cs:35]:
```suggestion
                    _messageBuilderFactory = ApplicationContext.GetService<IMessageBuilderFactory>() ?? new DefaultMessageBuilderFactory();
```

File: src/Integration/src/IntegrationBase/Util/IntegrationContextUtils.cs [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+            return context.GetService<IEvaluationContext>(INTEGRATION_SIMPLE_EVALUATION_CONTEXT_BEAN_NAME);
+        }
+
+        public static IExpressionParser GetExpressionParser(IApplicationContext context)
+        {
+            var result = context.GetService<IExpressionParser>();
+            if (result == null)
+            {
+                result = new ExpressionParser();
+            }
```
Comment (TimHess) [inline src/Integration/src/IntegrationBase/Util/IntegrationContextUtils.cs:48]:
```suggestion
            var result = context.GetService<IExpressionParser>() ?? new ExpressionParser();
```

File: src/Integration/src/IntegrationBase/Util/IntegrationServicesUtils.cs [hunk 9]
```diff
@@ -0,0 +1,49 @@
...
+            return ex;
+        }
+
+        public static IIntegrationServices GetIntegrationServices(IApplicationContext context)
+        {
+            var services = context?.GetService<IIntegrationServices>();
+            if (services == null)
+            {
+                services = new IntegrationServices(context);
+            }
```
Comment (TimHess) [inline src/Integration/src/IntegrationBase/Util/IntegrationServicesUtils.cs:44]:
```suggestion
            var services = context?.GetService<IIntegrationServices>() ?? new IntegrationServices(context);
```

File: src/Stream/src/StreamBase/Binder/PartitionHandler.cs [hunk 10]
```diff
@@ -35,15 +37,51 @@ public class PartitionHandler
 
         public int PartitionCount { get; set; }
 
+        public IExpressionParser ExpressionParser
+        {
+            get
+            {
+                if (_expressionParser != null)
```
Comment (TimHess) [inline src/Stream/src/StreamBase/Binder/PartitionHandler.cs:44]: Is this correct, or should it be `== null`?

File: src/Stream/src/StreamBase/Binder/PartitionHandler.cs [hunk 11]
```diff
@@ -35,15 +37,51 @@ public class PartitionHandler
...
+            {
+                _expressionParser = value;
+            }
+        }
+
+        public IEvaluationContext EvaluationContext
+        {
+            get
+            {
+                if (_evaluationContext != null)
```
Comment (TimHess) [inline src/Stream/src/StreamBase/Binder/PartitionHandler.cs:62]: Same question as above

File: src/Stream/src/StreamBase/Binder/PartitionHandler.cs (hunk 10 above)
Comment (dtillman) [reply src/Stream/src/StreamBase/Binder/PartitionHandler.cs]: Fixed

File: src/Stream/src/StreamBase/Binder/PartitionHandler.cs (hunk 11 above)
Comment (dtillman) [reply src/Stream/src/StreamBase/Binder/PartitionHandler.cs]: Fixed

**Pull request merged**